### PR TITLE
Add pickle test

### PR DESCRIPTION
**Pull request by tlambert03**
this includes #26  ... and modifies the get/set state accordingly to deal with the new lock object.

@VolkerH ... after #26, have a look here

### CODE REVIEW

**Comment by VolkerH** [comment]
Thanks!
 I will have a look first thing in the morning (my time).

File: tests/test_reader.py [hunk 1]
```diff
@@ -195,3 +195,23 @@ def _common_entries(*dcts):
...
+    f = ND2File(single_nd2)
+    pf = pickle.dumps(f)
+    assert isinstance(pf, bytes)
+    f2: ND2File = pickle.loads(pf)
+    np.testing.assert_array_equal(f, f2)
+
+    d = f.to_dask()
+    pd = pickle.dumps(d)
+    assert isinstance(pd, bytes)
+    # FIXME: this line currently leaks a file handle...
```
Comment (VolkerH) [inline tests/test_reader.py:212]: Is this due to the `__setstate__` method when unpickling?
`__setstate__` will open the file, but I don't see the mechanism (context manager?) that would close the file again for an unpickled array.  Maybe through garbage collection?

**Review by VolkerH — APPROVED** [review]
Hi @tlambert03,

Thanks for this.  I am awed. 

I pulled this PR branch and it seems to work for my use cases, I think everything I mentioned in issues #19 and #25.
Moreover, I can now pip install locally and the shared library is found, so it seems to address #24 as well.

I left a couple of comments, nothing I feel strongly about. Would be great to have this merged and rolled out.
Thanks again!

**Comment by VolkerH** [comment]
Ooops .... 
I was too quick. 
While my tests pass, my actual application still crashes when I return the dask array from a context manager. 
I will try and debug where that happens and provide a minimal example.

**Comment by VolkerH** [comment]
So what I see in the context of a larger application (shortened traceback) when returning the dask array from a ND2File context manager. When I leave ND2File open and return the array, I don't observe the crashes.


When I try and create a smaller example consisting of the following two steps

* call function that reads nd2 file and returns dask_array from context manager
* `np.array(da_returned_from_context_manager[0,0,:,:])`

it doesn't crash. Although these are exactly the steps that happen in the context of  the larger application. 

```
Fatal Python error: Segmentation fault

Thread 0x00007f26dba25700 (most recent call first):
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/concurrent/futures/thread.py", line 75 in _worker
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/threading.py", line 892 in run
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/threading.py", line 954 in _bootstrap_inner
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/threading.py", line 912 in _bootstrap
  File "/home/hilsenst/.vscode/extensions/ms-python.python-2021.11.1422169775/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydev_bundle/pydev_monkey.py", line 1054 in __call__

Thread 0x00007f275e7fc700 (most recent call first):
```

**Comment by VolkerH** [comment]
Now I can reproduce it with a small example

my `test_nd2.py`
```python

```
import dask.array as da
import numpy as np
from nd2 import ND2File

# TODO: create a small test .nd2 on the Nikon that we can add to resources
# or add something to git lfs
dataset_nd2 = "/home/hilsenst/Documents/Luisa_Reference_HT/PreMaldi/Seq0000.nd2"

# removed a number of tests here for brevity
...

def _load_nd2(filepath):
    with ND2File(filepath) as f:
        return f.to_dask()


def test_nd2_from_context_mgr():
    arr = _load_nd2(dataset_nd2)
    np_slice = np.array(arr[0, 0, :, :])

```
tests/test_nd2.py ....Fatal Python error: Segmentation fault

Thread 0x00007f4c0ecff700 (most recent call first):
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/concurrent/futures/thread.py", line 75 in _worker
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/threading.py", line 892 in run
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/threading.py", line 954 in _bootstrap_inner
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/threading.py", line 912 in _bootstrap

Current thread 0x00007f4c45f31180 (most recent call first):
  File "/home/hilsenst/GitlabEMBL/spacem-ht/src/spacem-mosaic/tests/test_nd2.py", line 53 in test_nd2_from_context_mgr
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/python.py", line 183 in pytest_pyfunc_call
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/python.py", line 1641 in runtest
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/runner.py", line 162 in pytest_runtest_call
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/runner.py", line 255 in <lambda>
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/runner.py", line 311 in from_call
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/runner.py", line 254 in call_runtest_hook
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/runner.py", line 215 in call_and_report
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/runner.py", line 126 in runtestprotocol
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/runner.py", line 109 in pytest_runtest_protocol
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/main.py", line 348 in pytest_runtestloop
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/main.py", line 323 in _main
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/main.py", line 269 in wrap_session
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/config/__init__.py", line 162 in main
  File "/home/hilsenst/miniconda3/envs/napari_latest/lib/python3.9/site-packages/_pytest/config/__init__.py", line 185 in console_main
  File "/home/hilsenst/miniconda3/envs/napari_latest/bin/pytest", line 8 in <module>
Segmentation fault (core dumped)

```

**Comment by tlambert03** [comment]
thank you @VolkerH!  I really appreciate the help.  I think we're narrowing in on something.

I will note that I found a few edge cases where I had to specify the scheduler to avoid a segfault (but I _usually_ was able to figure out why), so let me try your example here.

not that this is a nice solution, but perhaps try the following?

```python
def test_nd2_from_context_mgr():
    arr = _load_nd2(dataset_nd2)
    np_slice = np.array(arr[0, 0, :, :].compute(scheduler='threads'))
```

**Comment by tlambert03** [comment]
I'm inclined to merge what we have so far, and keep working at the edge cases... that ok with you?

**Comment by tlambert03** [comment]
> Now I can reproduce it with a small example
> 
> my `test_nd2.py`

hmmm, this test actually passes fine for (on mac).  you're on ubuntu yes?

**Comment by VolkerH** [comment]
Yes, Ubuntu 20.04.
I initially had some trouble reproducing this error with very similar code.
I'm not sure how much this is effected by some sort of external state
(number of cores being used, race conditions between threads, etc)

**Comment by VolkerH** [comment]
> I'm inclined to merge what we have so far, and keep working at the edge cases... that ok with you?

Yes, I'm all for it. 
I'm working with this version already and just work around the context manager issue by leaving the ND2File object in open state. Happy to leave this for a while, I think getting to the bottom of that threading/segfault problem might be time-consuming (and probably exceeds my skills)

**Comment by tlambert03** [comment]
got the leaked file figured out!  One of the key details was that the dask array wasn't being "unpickled" back to a the `ResourceBackedDaskArray` ... (but rather a regular dask array).  Overriding the `__reduce__` method let me restore it as the ResourceBacked array, wherein I could call setstate and close the underlying file_ctx.  learned a thing or two about pickle in the process!

**Comment by VolkerH** [comment]
Just saw this. Great that you could track it down.
Will test first thing Monday!
